src/qengine/opencl.cpp

```diff
@@ -550,15 +556,102 @@ void QEngineOCL::Decohere(bitLenInt start, bitLenInt length, QInterfacePtr desti
 
 void QEngineOCL::Dispose(bitLenInt start, bitLenInt length) { DecohereDispose(start, length, (QEngineOCLPtr) nullptr); }
 
+/// PSEUDO-QUANTUM Check whether bit phase is separable in permutation basis
+bool QEngineOCL::IsPhaseSeparable(bool forceCheck)
+{
+    if ((!forceCheck) && knowIsPhaseSeparable) {
+        return isPhaseSeparable;
+    }
+
+    if (doNormalize && (runningNorm != ONE_R1)) {
+        NormalizeState();
+    }
+
+    bitCapInt bciArgs[BCI_ARG_LEN] = { maxQPower, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+
+    queue.enqueueWriteBuffer(ulongBuffer, CL_FALSE, 0, sizeof(bitCapInt) * BCI_ARG_LEN, bciArgs);
+    queue.flush();
+
+    bitCapInt maxI = bciArgs[0];
+    size_t ngc = FixWorkItemCount(maxI, nrmGroupCount);
+    size_t ngs = FixGroupSize(ngc, nrmGroupSize);
+
+    bitLenInt* isAllSame = new bitLenInt[ngc];
+    std::fill(isAllSame, isAllSame + ngc, 1);
+    real1* phases = new real1[ngc];
+    std::fill(phases, phases + ngc, -PI_R1 * 2);
+
+    cl::Buffer isAllSameBuffer =
+        cl::Buffer(context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_WRITE, sizeof(bitLenInt) * ngc, isAllSame);
+    cl::Buffer phasesBuffer = cl::Buffer(context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_WRITE, sizeof(real1) * ngc, phases);
+
+    OCLDeviceCall ocl = device_context->Reserve(OCL_API_ISPHASESEPARABLE);
+    queue.finish();
+    ocl.call.setArg(0, *stateBuffer);
+    ocl.call.setArg(1, ulongBuffer);
+    ocl.call.setArg(2, phasesBuffer);
+    ocl.call.setArg(3, isAllSameBuffer);
+
+    // Note that the global size is 1 (serial). This is because the kernel is not very easily parallelized, but we
+    // ultimately want to offload all manipulation of stateVec from host code to OpenCL kernels.
+    queue.enqueueNDRangeKernel(ocl.call, cl::NullRange, // kernel, offset
+        cl::NDRange(ngc), // global number of work items
+        cl::NDRange(ngs)); // local number (per group)
+
+    bool toRet = true;
+    queue.enqueueMapBuffer(isAllSameBuffer, CL_TRUE, CL_MAP_READ, 0, sizeof(bitLenInt) * ngc);
+    for (size_t i = 0; i < ngc; i++) {
+        toRet &= (isAllSame[i] == 1);
+    }
+    queue.enqueueUnmapMemObject(isAllSameBuffer, isAllSame);
+    queue.finish();
+
+    delete[] isAllSame;
```

Comment:
Do cleanup in the same location as the other cleanup code, towards the end.  That way allocations and deallocations happen in visible pairs.

Reply:
Done.